llvm/lib/Target/AMDGPU/AMDGPUCombine.td

```diff
@@ -145,6 +145,31 @@ def expand_promoted_fmed3 : GICombineRule<
 
 } // End Predicates = [NotHasMed3_16]
 
+def promote_i16_uniform_binops_frag : GICombinePatFrag<
```

Comment:
I'll look at moving this into the legalizer

Reply:
I just remembered that the legalizer doesn't have uniformity info. This feels more like a legalization rule but it looks like it'll be stuck here due to the need for uniformity info, unless I'm missing another way to implement it?

Reply:
It's not really legalization. It's a register bank aware optimization. I would drop all the globalisel handling from this patch, it's a separate issue

Reply:
This will switch the CGP i16 promotion off, so we'd let GISel regress in the meantime?

Reply:
I'd first implement all the DAG parts in one patch and leave the IR pass as it is. Then the next 2 parts can be done in either order, either do the globalisel parts or remove the IR promotion and let the regression happen